src/test/rgw/test_d4n_filter.cc

```diff
@@ -0,0 +1,2819 @@
+    */
+
+    conn->cancel();
+    testBucket->remove(env->dpp, true, optional_yield{yield});
+    DriverDestructor driver_destructor(static_cast<rgw::sal::D4NFilterDriver*>(driver));
+  }, rethrow);
+
+  io.run();
+}
+
+// Read cache tests, versioned
+TEST_F(D4NFilterFixture, PutVersionedObjectRead)
+{
+  const std::string testName = "PutVersionedObjectRead";
+ 
+  boost::asio::spawn(io, [this, &testName] (boost::asio::yield_context yield) {
+    init_driver(yield);
+    create_bucket(testName, yield);
+    std::string instance;
+    put_version_enabled_object(testName, instance, yield);
+    put_version_suspended_object(testName, yield);
+
+    // TODO: No unversioned head object set in redis, is this intended?
+    boost::system::error_code ec;
+    request req;
+    req.push("EXISTS", TEST_BUCKET + testName + "__:" + instance + "_" + TEST_OBJ + testName + "_0_0");
+    req.push("EXISTS", TEST_BUCKET + testName + "__:null_" + TEST_OBJ + testName + "_0_0");
+    req.push("HGETALL", TEST_BUCKET + testName + "__:" + instance + "_" + TEST_OBJ + testName + "_0_0");
+    req.push("HGETALL", TEST_BUCKET + testName + "__:null_" + TEST_OBJ + testName + "_0_0");
+    req.push("FLUSHALL");
+
+    response< int, int, 
+              std::map<std::string, std::string>,
+              std::map<std::string, std::string>,
+	      boost::redis::ignore_t > resp;
+
+    conn->async_exec(req, resp, yield[ec]);
+
+    ASSERT_EQ((bool)ec, false);
+    EXPECT_EQ((int)std::get<0>(resp).value(), 1);
+    EXPECT_EQ((int)std::get<1>(resp).value(), 1);
+    EXPECT_EQ(std::get<2>(resp).value().size(), 10);
+    EXPECT_EQ(std::get<3>(resp).value().size(), 10);
+
+    conn->cancel();
+    testBucket->remove(env->dpp, true, optional_yield{yield});
+    DriverDestructor driver_destructor(static_cast<rgw::sal::D4NFilterDriver*>(driver));
+  }, rethrow);
+
+  io.run();
+}
+
+TEST_F(D4NFilterFixture, GetVersionedObjectRead)
+{
+  const std::string testName = "GetVersionedObjectRead";
+ 
+  boost::asio::spawn(io, [this, &testName] (boost::asio::yield_context yield) {
+    init_driver(yield);
+    create_bucket(testName, yield);
+    std::string instance;
+    put_version_enabled_object(testName, instance, yield);
+    put_version_suspended_object(testName, yield);
+
+    // For version enabled object
+    std::unique_ptr<rgw::sal::Object::ReadOp> read_op_enabled(objEnabled->get_read_op());
+    EXPECT_EQ(read_op_enabled->prepare(optional_yield{yield}, env->dpp), 0);
+    EXPECT_EQ(read_op_enabled->iterate(env->dpp, 0, ofs, nullptr, optional_yield{yield}), 0);
+
+    // For version suspended object
+    std::unique_ptr<rgw::sal::Object::ReadOp> read_op_suspended(objSuspended->get_read_op());
+    EXPECT_EQ(read_op_suspended->prepare(optional_yield{yield}, env->dpp), 0);
+    EXPECT_EQ(read_op_suspended->iterate(env->dpp, 0, ofs, nullptr, optional_yield{yield}), 0);
+
+    // TODO: data block set without version in key, is this intended?
```

Comment:
TODO:  Is it expected for objects with versioning suspended to not have their version ("null") in the directory entry key for the data block?

Reply:
if the version is "null" the oid is the name of the object, the oid does not include "null" as the version, so this is expected.